collect_app/src/main/java/org/odk/collect/android/openrosa/okhttp/OkHttpOpenRosaServerClientProvider.java

```diff
@@ -49,37 +52,47 @@ public class OkHttpOpenRosaServerClientProvider implements OpenRosaServerClientP
     private static final String DATE_HEADER = "Date";
 
     private final OkHttpClient baseClient;
+    private final String cacheDir;
 
-    private HttpCredentialsInterface lastCredentials;
-    private OkHttpOpenRosaServerClient client;
+    private final Map<HttpCredentialsInterface, OkHttpOpenRosaServerClient> clients = new HashMap<>();
 
     public OkHttpOpenRosaServerClientProvider(@NonNull OkHttpClient baseClient) {
+        this(baseClient, null);
+    }
+
+    public OkHttpOpenRosaServerClientProvider(@NonNull OkHttpClient baseClient, String cacheDir) {
         this.baseClient = baseClient;
+        this.cacheDir = cacheDir;
     }
 
     @Override
-    public OpenRosaServerClient get(String scheme, String userAgent, @Nullable HttpCredentialsInterface credentials) {
-        if (client == null || credentialsHaveChanged(credentials)) {
-            lastCredentials = credentials;
-            client = createNewClient(scheme, userAgent, credentials);
+    public synchronized OpenRosaServerClient get(String scheme, String userAgent, @NonNull HttpCredentialsInterface credentials) {
+        OkHttpOpenRosaServerClient existingClient = clients.get(credentials);
+
+        if (existingClient == null) {
+            OkHttpOpenRosaServerClient newClient = createNewClient(scheme, userAgent, credentials);
+            clients.put(credentials, newClient);
+            return newClient;
+        } else {
+            return existingClient;
         }
-
-        return client;
-    }
-
-    public boolean credentialsHaveChanged(@Nullable HttpCredentialsInterface credentials) {
-        return lastCredentials != null && !lastCredentials.equals(credentials)
-                || lastCredentials == null && credentials != null;
     }
 
     @NonNull
-    private OkHttpOpenRosaServerClient createNewClient(String scheme, String userAgent, @Nullable HttpCredentialsInterface credentials) {
+    private OkHttpOpenRosaServerClient createNewClient(String scheme, String userAgent, @NonNull HttpCredentialsInterface credentials) {
         OkHttpClient.Builder builder = baseClient.newBuilder()
                 .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
                 .writeTimeout(WRITE_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
                 .readTimeout(READ_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
                 .followRedirects(true);
 
+        if (cacheDir != null) {
+            builder.cache(new Cache(
+                    new File(cacheDir, "http_" + credentials.hashCode()),
```

Comment:
I'm using the hash code of credentials here instead of a value generated off the `OkHttp` client itself so that the same cache will get used for the same credentials across different processes of the app.

Reply:
Shouldn't we take scheme into accunt here too so that two project with the same username/password do not share the same dir?

Reply:
> so that two project with the same username/password do not share the same dir?

I don't **think** it's a problem. I discussed it [here](https://github.com/getodk/collect/pull/5440#discussion_r1105080391).